Revive the Dynamic Workflow Code from Old to New Flytekit version

samhita-alla opened the pull request:
Signed-off-by: samhita-alla <aallasamhita@gmail.com>

samhita-alla commented on legacy/demos/gaic-2020/demo/house_price_predictor.py:
```diff
@@ -2,86 +2,132 @@
 # The example has been borrowed from Sagemaker examples
```
Done. I removed the changes in legacy files.

wild-endeavor commented on cookbook/case_studies/house_price_prediction/README.rst:
```diff
@@ -0,0 +1,58 @@
+ - How the Flyte type-system works
+ - Saving the dataset into CSV files and loading them again to fit the model and later, generating predictions
+
+Code Walkthrough
```
We can delete everything after this section.

wild-endeavor commented on cookbook/case_studies/house_price_prediction/sandbox.config:
```diff
@@ -0,0 +1,7 @@
+
+[auth]
+assumable_iam_role=arn:aws:iam::173840052742:role/flytefunctionaltestsbatchworker-production-iad
+raw_output_data_prefix=s3://lyft-modelbuilder/cookbook
```
```suggestion
raw_output_data_prefix=s3://flyte-demo/raw_data
```

samhita-alla replied:
Done.

wild-endeavor commented on cookbook/case_studies/house_price_prediction/sandbox.config:
```diff
@@ -0,0 +1,7 @@
+python_venv=flytekit_venv
+
+[auth]
+assumable_iam_role=arn:aws:iam::173840052742:role/flytefunctionaltestsbatchworker-production-iad
```
Can we get rid of this line?  I don't think it's necessary.  Can you try running without this line?  If it doesn't work, can you try replacing this line with `kubernetes_service_account=default`

samhita-alla replied:
The code is working locally without assumable_iam_role.

kumare3 commented on cookbook/case_studies/house_price_prediction/house_price_predictor.py:
```diff
@@ -0,0 +1,253 @@
+    return _price
+# %%
+# Random data pertaining to a house:
+def gen_random_house() -> typing.List:
```
how about converting this to a pandas dataframe?

kumare3 replied:
cc @cosmicBboy

kumare3 replied:
basically we can collapse the 2 functions into one

samhita-alla replied:
`gen_random_house()` function is to generate the house details for one house, and the `gen_houses` function is to generate all the house details which internally calls the former function.

Is there a necessity to merge them?

samhita-alla replied:
Done.

kumare3 commented on cookbook/case_studies/house_price_prediction/house_price_predictor.py:
```diff
@@ -0,0 +1,253 @@
+# -----------------------------------------------
+# Train and validation datasets are directories (consisting of one CSV file) and test data is a CSV file
+@task(cache=True, cache_version="0.1", limits=Resources(mem="600Mi"))
+def generate_and_split_data(
```
can this return pandas dataframes itself? and simplify rather than converting to a csv?

samhita-alla replied:
Done.

kumare3 commented on cookbook/case_studies/house_price_prediction/house_price_predictor.py:
```diff
@@ -0,0 +1,253 @@
+) -> FlyteFile[typing.TypeVar("joblib.dat")]:
+    """
+    This function takes the given input features and their corresponding classes to train a XGBClassifier.
+    NOTE: We have simplified the algorithm for demo. Default hyper params & no validation dataset :P.
```
lets actually have a validation dataset?

samhita-alla replied:
Done.

kumare3 commented on cookbook/case_studies/house_price_prediction/house_price_predictor.py:
```diff
@@ -0,0 +1,253 @@
+    train, _, test = generate_and_split_data(
+        loc=LOCATIONS[0], number_of_houses=NUM_HOUSES_PER_LOCATION, seed=7
+    )
+    fit_task = fit(loc=LOCATIONS[0], train=train)
```
```suggestion
    model = fit(loc=LOCATIONS[0], train=train)
```

kumare3 commented:
@samhita-alla couple comments, This is a great start, but the code is still written in the old style. the new API should make the code simpler and nicer.
Can we also have a validation dataset and make it lesser verbose.

kumare3 commented:
@cosmicBboy we could also use Pandera here and use pandera's hypothesis integration to test it?

samhita-alla commented:
> @samhita-alla couple comments, This is a great start, but the code is still written in the old style. the new API should make the code simpler and nicer.
> Can we also have a validation dataset and make it lesser verbose.

Cleaned the code now. Added validation data (for evaluation) while training the model.

kumare3 commented on cookbook/case_studies/house_price_prediction/house_price_predictor.py:
```diff
@@ -0,0 +1,232 @@
+
+# %%
+# Splitting the Data into Train, Val, and Test Datasets:
+def split_data(
```
can we use a named output?
NamedTuple?

samhita-alla replied:
Observing an issue concerning "NamedTuple" (`Error:RestrictedTypeError(f"Transformer for type {self.python_type} is restricted currently"`)). Looking into it.

kumare3 commented on cookbook/case_studies/house_price_prediction/house_price_predictor.py:
```diff
@@ -0,0 +1,232 @@
+# Step 4: Task -- Generating & Splitting the Data
+# -----------------------------------------------
+#
+# Train, validation, and test datasets are DataFrames.
```
lets explain why we generated and split the data in one task?

samhita-alla replied:
I don't think there's a specific reason as to why that's done.

samhita-alla replied:
It's just that there are two helper functions that we've already defined, and hence, we are calling them in a single function. Do you want me to write this?

kumare3 commented on cookbook/case_studies/house_price_prediction/house_price_predictor.py:
```diff
@@ -0,0 +1,232 @@
+
+
+# %%
+# You can call the ``house_price_predictor_trainer`` workflow to print the predictions. However, our goal is to expand this to multiple regions. Thus, the tasks and helper functions defined here are called in the other Python script.
```
do you want to have a __main__ section here to show how to run this example

samhita-alla replied:
Done.

kumare3 commented on cookbook/case_studies/house_price_prediction/multiregion_house_price_predictor.py:
```diff
@@ -0,0 +1,158 @@
+# (A "Dynamic" Task (aka Workflow) spins up internal workflows)
+#
+# Serialize the XGBoost models using joblib and store the models in dat files.
+@dynamic(cache=True, cache_version="0.1", limits=Resources(mem="600Mi"))
```
I think we should combine the parallel fit and parallel predict into one @dynamic task like
I think this is more powerful!

```python
@dynamic
def parallel_fit_predict(...)
for ... in .....:
    m = fit(...)
    v = predict(m)
```
@wild-endeavor Wdyt?

kumare3 replied:
we can add it in the comments that we can alternatively have 2 parallel methods, but this basically makes it possible to parallelize an entire set of tasks?

wild-endeavor replied:
sure.

samhita-alla replied:
Done.

kumare3 replied:
Done what?

samhita-alla replied:
I merged the two dynamic tasks into one.

kumare3 commented:
Thank you @samhita-alla for the PR, I added a few more comments, lets address then then +1

wild-endeavor commented on cookbook/case_studies/house_price_prediction/multiregion_house_price_predictor.py:
```diff
@@ -0,0 +1,158 @@
+    test_sets = []
+    for loc in locations:
+        _train, _val, _test = generate_and_split_data(
+            loc=loc, number_of_houses=number_of_houses_per_location, seed=seed
```
Let's remove loc from the input list of the `generate_and_split_data` task.  It's not used.

samhita-alla replied:
Done.
